pkg/operatorstatus/operator_status.go

```diff
@@ -198,13 +200,133 @@ func NewClusterOperatorStatusCondition(conditionType configv1.ClusterStatusCondi
+		controllerConditions = append(controllerConditions, controllerCondition{MachineSyncControllerAvailableCondition, MachineSyncControllerDegradedCondition})
+		controllerConditions = append(controllerConditions, controllerCondition{MachineSetSyncControllerAvailableCondition, MachineSetSyncControllerDegradedCondition})
+	}
+
+	// Variables to store the conditions evaluation.
+	availableMsg, degradedMsg := []string{}, []string{}
+	anyAvailableMissing, anyDegradedMissing := false, false
+
+	// Evaluate each controller's conditions.
+	for _, conditions := range controllerConditions {
+		evaluateCondition(co, conditions.available, &anyAvailableMissing, &availableMsg, false)
+		evaluateCondition(co, conditions.degraded, &anyDegradedMissing, &degradedMsg, true)
+	}
+
+	availableCondition := newAvailableCondition(anyAvailableMissing, availableMsg, releaseVersion)
+	degradedCondition := newDegradedCondition(anyDegradedMissing, degradedMsg)
+	progressingCondition := newCondition(configv1.OperatorProgressing, configv1.ConditionFalse, "", "")
+
+	upgradeableStatus := configv1.ConditionUnknown
+
+	switch {
+	case availableCondition.Status == "True" && degradedCondition.Status == "False":
+		upgradeableStatus = configv1.ConditionTrue
+	case availableCondition.Status == "False" || degradedCondition.Status == "True":
+		upgradeableStatus = configv1.ConditionFalse
+	}
+
+	upgradeableCondition := newCondition(configv1.OperatorUpgradeable, upgradeableStatus, "", "")
+
+	return []configv1.ClusterOperatorStatusCondition{
+		availableCondition,
+		degradedCondition,
+		progressingCondition,
+		upgradeableCondition,
+	}
+}
+
+// evaluateCondition helps to evaluate conditions and update states.
+func evaluateCondition(co *configv1.ClusterOperator, conditionType configv1.ClusterStatusConditionType, anyMissing *bool, messages *[]string, isDegraded bool) {
+	condition := findCondition(co.Status.Conditions, conditionType)
+	if condition == nil {
+		*anyMissing = true
+		return
+	}
+
+	expectedStatus := configv1.ConditionFalse
+	if isDegraded {
+		expectedStatus = configv1.ConditionTrue
+	}
+
+	if condition.Status == expectedStatus {
+		statusString := fmt.Sprintf("%s=%s", conditionType, condition.Status)
+		*messages = append(*messages, statusString)
+	}
+}
+
+// newAvailableCondition helps to create the Available condition.
+func newAvailableCondition(anyMissing bool, messages []string, releaseVersion string) configv1.ClusterOperatorStatusCondition {
+	if anyMissing {
+		return newCondition(configv1.OperatorAvailable, configv1.ConditionUnknown, "", "")
+	}
+
+	if len(messages) == 0 {
+		return newCondition(configv1.OperatorAvailable, configv1.ConditionTrue, ReasonAsExpected, fmt.Sprintf("Cluster CAPI Operator is available at %s", releaseVersion))
+	}
+
+	return newCondition(configv1.OperatorAvailable, configv1.ConditionFalse, "ControllersNotAvailable", fmt.Sprintf("The following controllers available conditions are not as expected: %s", strings.Join(messages, ", ")))
```

Comment:
As far as I could tell from reviewing the other controllers, we will never report available false, which I think is probably an error right?